In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

from numpy import mean
from numpy import std
from matplotlib import pyplot
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.preprocessing import QuantileTransformer
from sklearn.pipeline import Pipeline

In [ ]:
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [ ]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [ ]:
!pip install mne

     |████████████████████████████████| 7.5 MB 12.0 MB/s 


In [ ]:
#import pyeeg
from tensorflow import keras 

In [ ]:
%matplotlib inline

import glob
import math
import ntpath
import random, os, sys
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt
import seaborn as sns
#import edfreader
#import models
import mne
#import pyeeg
import warnings
warnings.filterwarnings('ignore')

from matplotlib.backends.backend_pdf import PdfPages
from mne.datasets.sleep_physionet._utils import _fetch_one, _data_path, AGE_SLEEP_RECORDS, _check_subjects
from datetime import datetime
from mne import Epochs, pick_types, find_events
from mne.io import concatenate_raws, read_raw_edf
from mne.time_frequency import psd_welch
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

from tqdm.notebook import tqdm
from sklearn.model_selection import train_test_split
from tensorflow import keras

from tensorflow.keras.utils import to_categorical

from tensorflow.keras.callbacks import ModelCheckpoint, ReduceLROnPlateau, CSVLogger
import imblearn
from imblearn.over_sampling import SMOTE
from sklearn.svm import SVC
import joblib

from sklearn.metrics import make_scorer, f1_score, accuracy_score, classification_report, log_loss
from sklearn.metrics import roc_auc_score, confusion_matrix, roc_auc_score, roc_curve

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
#from xgboost import XGBClassifier
#import xgboost

In [ ]:
VBS = True  # constant boolean to enable/disbale verbose
EPOCH_SEC_SIZE = 30  # Epoch duration selection
seed = 42  # seed value for the random seeds
batch_size = 64
number_of_files = 100 #100 npz files, 50 subjects, each suject has 2 nights, so 2 files per subject (50*2)=100

# values to label the stages
UNKNOWN = -1
W = 0
N1 = 1
N2 = 2
N3 = 3
REM = 4

# making string dictionary for the label values
label_dict = {
    "UNKNOWN"  : UNKNOWN,
    "W"        : W,
    "N1"       : N1,
    "N2"       : N2,
    "N3"       : N3,
    "REM"      : REM
}

# converting from label values to strings 
class_dict = {
    -1: "UNKNOWN",
    0 : "W",
    1 : "N1",
    2 : "N2",
    3 : "N3",
    4 : "REM"
}

# annotation dictionary to convert from string to label values
annot2label = {
    "Sleep stage ?": -1,
    "Movement time": -1,
    "Sleep stage W": 0,
    "Sleep stage 1": 1,
    "Sleep stage 2": 2,
    "Sleep stage 3": 3,
    "Sleep stage 4": 3,
    "Sleep stage R": 4
}
#project_path = os.path.abspath(os.getcwd())  # finding the current project path in windows

In [ ]:
# localized fetch_data function by using mne library
# https://github.com/mne-tools/mne-python/blob/maint/0.20/mne/datasets/sleep_physionet/age.py#L18-L111
data_path = _data_path
BASE_URL = 'https://physionet.org/physiobank/database/sleep-edfx/sleep-cassette/'
def fetch_data(subjects, recording=[1, 2], path=None, force_update=False,
               update_path=None, base_url=BASE_URL,
               verbose=None):  # noqa: D301
    records = np.loadtxt(AGE_SLEEP_RECORDS,
                         skiprows=1,
                         delimiter=',',
                         usecols=(0, 1, 2, 6, 7),
                         dtype={'names': ('subject', 'record', 'type', 'sha',
                                          'fname'),
                                'formats': ('<i2', 'i1', '<S9', 'S40', '<S22')}
                         )
    psg_records = records[np.where(records['type'] == b'PSG')]
    hyp_records = records[np.where(records['type'] == b'Hypnogram')]

    path = data_path(path=path, update_path=update_path)
    params = [path, force_update, base_url]
    fnames = []
    for subject in subjects:
        for idx in np.where(psg_records['subject'] == subject)[0]:
            if psg_records['record'][idx] in recording:
                psg_fname = _fetch_one(psg_records['fname'][idx].decode(),
                                       psg_records['sha'][idx].decode(),
                                       *params)
                hyp_fname = _fetch_one(hyp_records['fname'][idx].decode(),
                                       hyp_records['sha'][idx].decode(),
                                       *params)
                fnames.append([psg_fname, hyp_fname])

    return fnames

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')
#%cd /gdrive

Mounted at /content/gdrive


In [ ]:
npz_files = sorted(glob.glob("/gdrive/My Drive/NPZ_files/***.npz"))
X = np.zeros((0, 3000, 1))
y = []
for fn in tqdm(npz_files[:number_of_files]):
    samples = np.load(fn)
    X_data = samples['x']
    X = np.concatenate((X, X_data), axis=0)
    y.extend(samples['y'])
y = np.array(y)

#sorted(glob.glob("/gdrive/My Drive/NPZ_files/***.npz"))


  0%|          | 0/100 [00:00<?, ?it/s]

In [ ]:
npz_files

['/gdrive/My Drive/NPZ_files/SC4001E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4002E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4011E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4012E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4021E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4022E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4031E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4032E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4041E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4042E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4051E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4052E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4061E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4062E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4071E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4072E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4081E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4082E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4091E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4092E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4101E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4102E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4

In [ ]:
len(npz_files[:number_of_files])

100

In [ ]:
npz_files[:number_of_files]

['/gdrive/My Drive/NPZ_files/SC4001E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4002E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4011E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4012E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4021E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4022E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4031E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4032E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4041E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4042E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4051E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4052E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4061E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4062E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4071E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4072E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4081E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4082E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4091E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4092E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4101E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4102E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4

In [ ]:
npz_files[148:150]

['/gdrive/My Drive/NPZ_files/SC4821G0.npz',
 '/gdrive/My Drive/NPZ_files/SC4822G0.npz']

In [ ]:
X

array([[[ 8.11135578e+00],
        [ 1.74886456e+01],
        [ 2.12395611e+01],
        ...,
        [-1.03619051e+01],
        [-1.11120882e+01],
        [-2.10989022e+00]],

       [[-1.07369967e+01],
        [-1.13934069e+01],
        [-4.45421267e+00],
        ...,
        [ 5.88424911e+01],
        [ 4.83399277e+01],
        [ 5.36849823e+01]],

       [[ 6.13743591e+01],
        [ 3.86813202e+01],
        [ 4.95589752e+01],
        ...,
        [ 3.36175842e+01],
        [ 3.43677673e+01],
        [ 3.18358974e+01]],

       ...,

       [[ 3.47106218e+00],
        [-5.03052510e-02],
        [ 2.56556773e+00],
        ...,
        [ 9.55799758e-01],
        [-1.86129427e+00],
        [ 3.37045169e+00]],

       [[ 2.26373625e+00],
        [ 3.52136761e-01],
        [ 2.86739922e+00],
        ...,
        [ 2.54041519e+01],
        [ 2.02730160e+01],
        [ 1.82608051e+01]],

       [[ 1.96693535e+01],
        [ 1.65504265e+01],
        [ 1.54437122e+01],
        ...,
        

In [ ]:
y

array([0, 0, 0, ..., 0, 0, 0], dtype=int32)

In [ ]:
from keras.utils import np_utils

In [ ]:
if VBS:
    print("Shape of the input data: {}".format(X.shape))
    print("Shape of the sleep stages: {}".format(y.shape))


Shape of the input data: (116718, 3000, 1)
Shape of the sleep stages: (116718,)


In [ ]:
from tensorflow.keras import optimizers, losses, activations, models
from tensorflow.keras.models import Model, load_model
from tensorflow.keras.layers import GlobalMaxPool1D, GlobalAveragePooling1D, concatenate, SpatialDropout1D, TimeDistributed, Bidirectional, LSTM
from tensorflow.keras.layers import Input, Conv1D, Dense, Dropout, MaxPool1D, Activation
from tensorflow.keras.layers import Reshape, LSTM, TimeDistributed, Bidirectional, BatchNormalization, Flatten, RepeatVector
from tensorflow.keras.optimizers import Adam
#from keras_contrib.layers import CRF

from scipy.signal import butter, lfilter


In [ ]:
#n_classes=5
def model_b(n_classes=5, use_sub_layer=False, use_rnn=True, verbose=False):
    inputLayer = Input(shape=(3000, 1), name='inLayer')
    convFine = Conv1D(filters=64, kernel_size=int(Fs/2), strides=int(Fs/16), padding='same', activation='relu', name='fConv1')(inputLayer)
    convFine = MaxPool1D(pool_size=8, strides=8, name='fMaxP1')(convFine)
    convFine = Dropout(rate=0.5, name='fDrop1')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv2')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv3')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv4')(convFine)
    convFine = MaxPool1D(pool_size=4, strides=4, name='fMaxP2')(convFine)
    fineShape = convFine.get_shape()
    convFine = Flatten(name='fFlat1')(convFine)
    
    # network to learn coarse features
    convCoarse = Conv1D(filters=32, kernel_size=Fs*4, strides=int(Fs/2), padding='same', activation='relu', name='cConv1')(inputLayer)
    convCoarse = MaxPool1D(pool_size=4, strides=4, name='cMaxP1')(convCoarse)
    convCoarse = Dropout(rate=0.5, name='cDrop1')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv2')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv3')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv4')(convCoarse)
    convCoarse = MaxPool1D(pool_size=2, strides=2, name='cMaxP2')(convCoarse)
    coarseShape = convCoarse.get_shape()
    convCoarse = Flatten(name='cFlat1')(convCoarse)
    
    # concatenate coarse and fine cnns
    mergeLayer = concatenate([convFine, convCoarse], name='merge_1')
    outLayer = Dropout(rate=0.5, name='mDrop1')(mergeLayer)
    
    outLayer = Reshape((1, outLayer.get_shape()[1]), name='reshape1')(outLayer)
    outLayer = LSTM(64, return_sequences=True)(outLayer)
    outLayer = LSTM(64, return_sequences=False)(outLayer)

    # Classify
    outLayer = Dense(n_classes, activation='softmax', name='outLayer')(outLayer)
    model = Model(inputLayer, outLayer)
    optimizer = Adam(lr=1e-4)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['acc'])
    #model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    if verbose:
        model.summary()
    return model

In [ ]:
def butter_bandpass(lowcut, highpass, fs, order=4):
    nyq = 0.5 * fs
    #       low = lowcut / nyq
    high = highpass / nyq
    b, a = butter(order, high, btype='highpass')
    return b, a

In [ ]:
def butter_bandpass_filter(data, highpass, fs, order=4):
    b, a = butter_bandpass(0, highpass, fs, order=order)
    y = lfilter(b, a, data)
    return y

In [ ]:
#npz_files = sorted(glob.glob(os.path.join(output_path, "*.npz")))
X_sub_482 = np.zeros((0, 3000, 1))
y_sub_482 = []
for fn in tqdm(npz_files[148:150]): #extracting data from the npz files corresponding to both tge recording 1, 2 of the 50th subject (no 53rd subject)
    samples = np.load(fn)
    X_data = samples['x']
    X_sub_482 = np.concatenate((X_sub_482, X_data), axis=0)
    y_sub_482.extend(samples['y'])
y_sub_482 = np.array(y_sub_482)

  0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
X_sub_482

array([[[-49.10647202],
        [-48.57606888],
        [-45.21685028],
        ...,
        [  1.81220996],
        [ -2.51941395],
        [  3.75702071]],

       [[ 17.63589668],
        [  4.37582397],
        [ 25.59194183],
        ...,
        [ 52.20048904],
        [ 55.55970764],
        [ 57.06251526]],

       [[ 50.25567627],
        [ 55.64810562],
        [ 50.16727829],
        ...,
        [-13.03907204],
        [-16.75189209],
        [-13.12747288]],

       ...,

       [[  4.04444456],
        [  2.26666665],
        [  6.44444466],
        ...,
        [ -6.17777777],
        [ 10.17777824],
        [-48.93333435]],

       [[-43.95555496],
        [-63.42222214],
        [-70.8888855 ],
        ...,
        [ 20.66666603],
        [ 14.88888931],
        [ 22.35555649]],

       [[ 21.73333359],
        [ 21.55555534],
        [ 21.91111183],
        ...,
        [  1.28888893],
        [  2.97777772],
        [  5.28888893]]])

In [ ]:
y_sub_482

array([0, 0, 0, ..., 0, 0, 0], dtype=int32)

In [ ]:
if VBS:
    print("Shape of the subject(482) data: {}".format(X_sub_482.shape))
    print("Shape of the corresponding sleep stage: {}".format(y_sub_482.shape))

Shape of the subject(482) data: (3070, 3000, 1)
Shape of the corresponding sleep stage: (3070,)


In [ ]:
pp_x = np.array([butter_bandpass_filter(sample, highpass=40.0, fs=100, order=4) for sample in X])

if VBS:
    print(pp_x.shape)

(116718, 3000, 1)


In [ ]:
y.shape

(116718,)

In [ ]:
type(y)

numpy.ndarray

In [ ]:
pp_x

array([[[ 3.91319654e-02],
        [ 8.43712310e-02],
        [ 1.02466935e-01],
        ...,
        [-4.99893880e-02],
        [-5.36085289e-02],
        [-1.01788349e-02]],

       [[-5.17989585e-02],
        [-5.49657067e-02],
        [-2.14886513e-02],
        ...,
        [ 2.83876381e-01],
        [ 2.33208409e-01],
        [ 2.58994788e-01]],

       [[ 2.96090982e-01],
        [ 1.86611970e-01],
        [ 2.39089513e-01],
        ...,
        [ 1.62182769e-01],
        [ 1.65801910e-01],
        [ 1.53587300e-01]],

       ...,

       [[ 1.67455958e-02],
        [-2.42689804e-04],
        [ 1.23771796e-02],
        ...,
        [ 4.61110622e-03],
        [-8.97952265e-03],
        [ 1.62602162e-02]],

       [[ 1.09210409e-02],
        [ 1.69882864e-03],
        [ 1.38333184e-02],
        ...,
        [ 1.22558352e-01],
        [ 9.78039900e-02],
        [ 8.80963939e-02]],

       [[ 9.48917148e-02],
        [ 7.98449401e-02],
        [ 7.45057705e-02],
        ...,
        

In [ ]:
pp_x_rshp=pp_x.reshape(116718, 3000)

In [ ]:
pp_x_rshp_df=pd.DataFrame(pp_x_rshp)

In [ ]:
pp_x_rshp_df['sleep_stage']=y

In [ ]:
pp_x_rshp_df

0         1         2  ...      2998      2999  sleep_stage
0       0.039132  0.084371  0.102467  ... -0.053609 -0.010179            0
1      -0.051799 -0.054966 -0.021489  ...  0.233208  0.258995            0
2       0.296091  0.186612  0.239090  ...  0.165802  0.153587            0
3       0.134134  0.116491  0.116491  ...  0.205612  0.274829            0
4       0.333640  0.359878  0.381141  ...  0.122372  0.207422            0
...          ...       ...       ...  ...       ...       ...          ...
116713  0.117219  0.233710  0.088582  ... -0.099746 -0.122073            0
116714 -0.087611 -0.076447 -0.056061  ...  0.046839  0.049266            0
116715  0.016746 -0.000243  0.012377  ... -0.008980  0.016260            0
116716  0.010921  0.001699  0.013833  ...  0.097804  0.088096            0
116717  0.094892  0.079845  0.074506  ...  0.143430  0.085184            0

[116718 rows x 3001 columns]

In [ ]:
pp_x_rshp_df=pd.read_csv('/content/gdrive/My Drive/physionet/ensemble/50_subs/original_50_subs_dt_original_y.csv',header=None)

In [ ]:
pp_x_rshp_df

0         1         2         3         4         5         6     \
0       0.039132  0.084371  0.102467  0.033251  0.045465  0.026465  0.008369   
1      -0.051799 -0.054966 -0.021489 -0.001131 -0.016512 -0.114229 -0.056775   
2       0.296091  0.186612  0.239090  0.259900  0.302877  0.316901  0.272567   
3       0.134134  0.116491  0.116491  0.070347  0.066728  0.037322  0.030989   
4       0.333640  0.359878  0.381141  0.386117  0.373450  0.372998  0.360783   
...          ...       ...       ...       ...       ...       ...       ...   
116713  0.117219  0.233710  0.088582  0.195365  0.141488  0.136634  0.103629   
116714 -0.087611 -0.076447 -0.056061 -0.058488 -0.075962 -0.053149 -0.065769   
116715  0.016746 -0.000243  0.012377  0.020629  0.014319 -0.042956 -0.085669   
116716  0.010921  0.001699  0.013833 -0.000243  0.008494 -0.000243  0.002670   
116717  0.094892  0.079845  0.074506  0.070623  0.077903  0.084213  0.082757   

            7         8         9     ...      2991      2992      2993  \
0       0.001131 -0.006560  0.009274  ... -0.034156  0.001583 -0.057228   
1      -0.085728 -0.113324 -0.078490  ...  0.264876  0.197922  0.267138   
2       0.281614  0.220541  0.265328  ...  0.164445  0.146349  0.176207   
3       0.039584  0.023751  0.020131  ... -0.087086 -0.040942  0.009726   
4       0.367117  0.359878  0.382046  ... -0.121920 -0.060847  0.000679   
...          ...       ...       ...  ...       ...       ...       ...   
116713  0.213810  0.167699  0.141488  ... -0.108482 -0.110424 -0.112365   
116714 -0.104114 -0.086155 -0.121102  ...  0.034219  0.093921  0.097319   
116715 -0.076447 -0.066740 -0.060915  ...  0.031792  0.027909  0.014804   
116716 -0.001699 -0.003640  0.011406  ...  0.126927  0.115763  0.124014   
116717  0.076447  0.084213  0.090523  ...  0.198278  0.121588  0.224488   

            2994      2995      2996      2997      2998      2999  3000  
0      -0.041394 -0.023751  0.002488 -0.049989 -0.053609 -0.010179     0  
1       0.264876  0.312377  0.251304  0.283876  0.233208  0.258995     0  
2       0.176659  0.149516  0.163540  0.162183  0.165802  0.153587     0  
3       0.035060  0.129610  0.045465  0.126896  0.205612  0.274829     0  
4       0.060394  0.045465  0.084371  0.057680  0.122372  0.207422     0  
...          ...       ...       ...       ...       ...       ...   ...  
116713 -0.064798 -0.171096 -0.115278 -0.044898 -0.099746 -0.122073     0  
116714  0.070623  0.069167  0.040529  0.058488  0.046839  0.049266     0  
116715  0.017231  0.020143  0.001213  0.004611 -0.008980  0.016260     0  
116716  0.130810  0.130810  0.124014  0.122558  0.097804  0.088096     0  
116717  0.146342  0.061886  0.129354  0.088582  0.143430  0.085184     0  

[116718 rows x 3001 columns]

In [ ]:
dt_shuf1=pp_x_rshp_df.sample(frac=1,replace=True, random_state=991)

In [ ]:
dt_shuf1

0         1         2         3         4         5         6     \
87982  -0.088735 -0.104907 -0.132443 -0.152111 -0.109278 -0.072563 -0.058577   
68528   0.600975  0.594189  0.584011  0.579345  0.569166  0.521241  0.527603   
37536  -0.034046 -0.019160 -0.031151  0.006478 -0.041076  0.005238 -0.006754   
49751  -0.057419  0.039738 -0.062918  0.072734 -0.031296  0.052111 -0.072084   
55894  -0.095076 -0.096344 -0.092538 -0.090000 -0.103957 -0.093807 -0.101843   
...          ...       ...       ...       ...       ...       ...       ...   
66766   0.052937  0.026246  0.023744  0.017488 -0.033809 -0.036312 -0.043818   
42895  -0.049391  0.001550 -0.088372 -0.002879 -0.073754 -0.010410 -0.087486   
7912   -0.019850 -0.035156  0.010284  0.041374  0.065768  0.088248  0.081552   
56020   0.194850  0.157920  0.165215  0.203969  0.175246  0.200778  0.200322   
114355  0.089229 -0.028979  0.099336 -0.037768  0.094062 -0.079514  0.078243   

            7         8         9     ...      2991      2992      2993  \
87982  -0.048961 -0.083053 -0.076060  ... -0.004816  0.005674 -0.010935   
68528   0.526331  0.537782  0.505549  ...  0.162437  0.172616  0.149290   
37536  -0.003032 -0.011716  0.016816  ...  0.079671  0.063543  0.043695   
49751   0.065401 -0.058335  0.064943  ... -0.210943 -0.027172 -0.199945   
55894  -0.109878 -0.104380 -0.090423  ...  0.042803  0.041534  0.038151   
...          ...       ...       ...  ...       ...       ...       ...   
66766  -0.009620 -0.005033 -0.022966  ... -0.022132  0.001640  0.007062   
42895  -0.012182 -0.066667 -0.031229  ...  0.040975  0.017054  0.055592   
7912    0.098771  0.106902  0.093988  ...  0.086813  0.081552  0.071507   
56020   0.205337  0.195306  0.163392  ... -0.005757  0.061720  0.006553   
114355 -0.074680  0.065939 -0.110274  ...  0.025511  0.023753 -0.003931   

            2994      2995      2996      2997      2998      2999  3000  
87982  -0.013995 -0.031915 -0.002631  0.027965  0.047633  0.040203     3  
68528   0.107302  0.092034  0.062345  0.065738  0.092458  0.064042     0  
37536   0.073881  0.109444  0.090835  0.099106  0.117714  0.104481     0  
49751  -0.016631 -0.211860  0.001242 -0.197653 -0.003799 -0.241190     0  
55894   0.025039  0.038151  0.045764  0.062681  0.065219  0.076638     1  
...          ...       ...       ...       ...       ...       ...   ...  
66766   0.003308  0.052103  0.062529  0.040008  0.028748  0.070036     4  
42895   0.012182  0.050277  0.011296  0.059136 -0.002879  0.044518     2  
7912    0.031808  0.052375  0.053810  0.091118  0.068638  0.093031     2  
56020   0.101385 -0.037216  0.051233 -0.003478  0.004273  0.003817     0  
114355 -0.001295 -0.010962 -0.028100 -0.036010 -0.023706 -0.014917     0  

[116718 rows x 3001 columns]

In [ ]:
dt_shuf1['sleep_stage']

87982     3
68528     0
37536     0
49751     0
55894     1
         ..
66766     4
42895     2
7912      2
56020     0
114355    0
Name: 3000, Length: 116718, dtype: int64

In [ ]:
x_shuf1=dt_shuf1.drop('sleep_stage',axis=1)
y_shuf1=dt_shuf1['sleep_stage']

In [ ]:
x_shuf1=dt_shuf1.drop([3000],axis=1)
y_shuf1=dt_shuf1[3000]

In [ ]:
x_shuf1

0         1         2         3         4         5         6     \
87982  -0.088735 -0.104907 -0.132443 -0.152111 -0.109278 -0.072563 -0.058577   
68528   0.600975  0.594189  0.584011  0.579345  0.569166  0.521241  0.527603   
37536  -0.034046 -0.019160 -0.031151  0.006478 -0.041076  0.005238 -0.006754   
49751  -0.057419  0.039738 -0.062918  0.072734 -0.031296  0.052111 -0.072084   
55894  -0.095076 -0.096344 -0.092538 -0.090000 -0.103957 -0.093807 -0.101843   
...          ...       ...       ...       ...       ...       ...       ...   
66766   0.052937  0.026246  0.023744  0.017488 -0.033809 -0.036312 -0.043818   
42895  -0.049391  0.001550 -0.088372 -0.002879 -0.073754 -0.010410 -0.087486   
7912   -0.019850 -0.035156  0.010284  0.041374  0.065768  0.088248  0.081552   
56020   0.194850  0.157920  0.165215  0.203969  0.175246  0.200778  0.200322   
114355  0.089229 -0.028979  0.099336 -0.037768  0.094062 -0.079514  0.078243   

            7         8         9     ...      2990      2991      2992  \
87982  -0.048961 -0.083053 -0.076060  ...  0.015289 -0.004816  0.005674   
68528   0.526331  0.537782  0.505549  ...  0.208666  0.162437  0.172616   
37536  -0.003032 -0.011716  0.016816  ...  0.088768  0.079671  0.063543   
49751   0.065401 -0.058335  0.064943  ... -0.022131 -0.210943 -0.027172   
55894  -0.109878 -0.104380 -0.090423  ...  0.019541  0.042803  0.041534   
...          ...       ...       ...  ...       ...       ...       ...   
66766  -0.009620 -0.005033 -0.022966  ...  0.001223 -0.022132  0.001640   
42895  -0.012182 -0.066667 -0.031229  ...  0.022370  0.040975  0.017054   
7912    0.098771  0.106902  0.093988  ...  0.054767  0.086813  0.081552   
56020   0.205337  0.195306  0.163392  ... -0.011684 -0.005757  0.061720   
114355 -0.074680  0.065939 -0.110274  ...  0.041330  0.025511  0.023753   

            2993      2994      2995      2996      2997      2998      2999  
87982  -0.010935 -0.013995 -0.031915 -0.002631  0.027965  0.047633  0.040203  
68528   0.149290  0.107302  0.092034  0.062345  0.065738  0.092458  0.064042  
37536   0.043695  0.073881  0.109444  0.090835  0.099106  0.117714  0.104481  
49751  -0.199945 -0.016631 -0.211860  0.001242 -0.197653 -0.003799 -0.241190  
55894   0.038151  0.025039  0.038151  0.045764  0.062681  0.065219  0.076638  
...          ...       ...       ...       ...       ...       ...       ...  
66766   0.007062  0.003308  0.052103  0.062529  0.040008  0.028748  0.070036  
42895   0.055592  0.012182  0.050277  0.011296  0.059136 -0.002879  0.044518  
7912    0.071507  0.031808  0.052375  0.053810  0.091118  0.068638  0.093031  
56020   0.006553  0.101385 -0.037216  0.051233 -0.003478  0.004273  0.003817  
114355 -0.003931 -0.001295 -0.010962 -0.028100 -0.036010 -0.023706 -0.014917  

[116718 rows x 3000 columns]

In [ ]:
y_shuf1_=np_utils.to_categorical(y_shuf1)

In [ ]:
x_shuf1_rshp=np.array(x_shuf1).reshape(116718,3000,1)

In [ ]:
pp_X_sub_482 = np.array([butter_bandpass_filter(sample, highpass=40.0, fs=100, order=4) for sample in X_sub_482])

if VBS:
    print(pp_X_sub_482.shape)

(3070, 3000, 1)


In [ ]:
pp_X_sub_482

array([[[-0.23690648],
        [-0.23434764],
        [-0.21814161],
        ...,
        [ 0.00874272],
        [-0.01215452],
        [ 0.01812516]],

       [[ 0.08508162],
        [ 0.02111048],
        [ 0.12346431],
        ...,
        [ 0.25183308],
        [ 0.26803911],
        [ 0.27528917]],

       [[ 0.24245064],
        [ 0.26846557],
        [ 0.24202418],
        ...,
        [-0.06290496],
        [-0.08081688],
        [-0.06333144]],

       ...,

       [[ 0.01951179],
        [ 0.01093518],
        [ 0.03109021],
        ...,
        [-0.02980372],
        [ 0.0491011 ],
        [-0.23607121]],

       [[-0.21205669],
        [-0.30597058],
        [-0.34199232],
        ...,
        [ 0.09970309],
        [ 0.07182911],
        [ 0.10785088]],

       [[ 0.10484906],
        [ 0.1039914 ],
        [ 0.10570673],
        ...,
        [ 0.00621804],
        [ 0.01436582],
        [ 0.02551542]]])

The following corresponds to reshaping the existing 3d data to 2d and converting it to .csv. The csv file was created on google drive and later downloaded to the local pc

In [ ]:
pp_x_rshp=pp_x.reshape(189560,3000)

In [ ]:
pp_x_rshp

array([[ 0.03913197,  0.08437123,  0.10246694, ..., -0.04998939,
        -0.05360853, -0.01017883],
       [-0.05179896, -0.05496571, -0.02148865, ...,  0.28387638,
         0.23320841,  0.25899479],
       [ 0.29609098,  0.18661197,  0.23908951, ...,  0.16218277,
         0.16580191,  0.1535873 ],
       ...,
       [-0.03455384,  0.00189911,  0.05185315, ..., -0.0525553 ,
         0.025301  ,  0.01360005],
       [-0.00620155,  0.02170071, -0.06515632, ...,  0.0352018 ,
         0.0203506 ,  0.0437525 ],
       [ 0.12205883, -0.08855822, -0.07505712, ..., -0.06470629,
        -0.08810818, -0.10835982]])

In [ ]:
pp_X_sub_482_rshp=pp_X_sub_482.reshape(3070,3000)

In [ ]:
pp_X_sub_482_rshp_df=pd.DataFrame(pp_X_sub_482_rshp)

In [ ]:
pp_X_sub_482_rshp_df

0         1         2     ...      2997      2998      2999
0    -0.236906 -0.234348 -0.218142  ...  0.008743 -0.012155  0.018125
1     0.085082  0.021110  0.123464  ...  0.251833  0.268039  0.275289
2     0.242451  0.268466  0.242024  ... -0.062905 -0.080817 -0.063331
3    -0.134126 -0.107258 -0.096170  ... -0.081670 -0.126023 -0.032199
4     0.009596  0.047125  0.021963  ...  0.048405 -0.019405 -0.001066
...        ...       ...       ...  ...       ...       ...       ...
3065  0.237787  0.278525  0.310259  ... -0.387877 -0.398598 -0.360432
3066 -0.405459 -0.413607 -0.363863  ...  0.086409  0.078262  0.086838
3067  0.019512  0.010935  0.031090  ... -0.029804  0.049101 -0.236071
3068 -0.212057 -0.305971 -0.341992  ...  0.099703  0.071829  0.107851
3069  0.104849  0.103991  0.105707  ...  0.006218  0.014366  0.025515

[3070 rows x 3000 columns]

In [ ]:
pp_X_sub_482_rshp_df['sleep_stage']=y_sub_482

In [ ]:
pp_X_sub_482_rshp_df

0         1         2  ...      2998      2999  sleep_stage
0    -0.236906 -0.234348 -0.218142  ... -0.012155  0.018125            0
1     0.085082  0.021110  0.123464  ...  0.268039  0.275289            0
2     0.242451  0.268466  0.242024  ... -0.080817 -0.063331            0
3    -0.134126 -0.107258 -0.096170  ... -0.126023 -0.032199            0
4     0.009596  0.047125  0.021963  ... -0.019405 -0.001066            0
...        ...       ...       ...  ...       ...       ...          ...
3065  0.237787  0.278525  0.310259  ... -0.398598 -0.360432            0
3066 -0.405459 -0.413607 -0.363863  ...  0.078262  0.086838            0
3067  0.019512  0.010935  0.031090  ...  0.049101 -0.236071            0
3068 -0.212057 -0.305971 -0.341992  ...  0.071829  0.107851            0
3069  0.104849  0.103991  0.105707  ...  0.014366  0.025515            0

[3070 rows x 3001 columns]

In [ ]:
pp_X_sub_482_rshp_df.to_csv(r'/gdrive/My Drive/filtered_x_sub_482.csv', header=False, index=False)

In [ ]:
Fs=100
#batch_size=10

30 epochs

In [ ]:
checkpoint_path_erfan_model = "/gdrive/My Drive/physionet/dt_shuf1_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf1_30_epochs.ckpt"

In [ ]:
checkpoint_path_erfan_model_dir = os.path.dirname(checkpoint_path_erfan_model)

In [ ]:
cp_callback_erfan_model_30_epochs = tf.keras.callbacks.ModelCheckpoint(
    filepath=checkpoint_path_erfan_model,
    monitor='accuracy', 
    verbose=1, 
    save_weights_only=True,
    mode = 'max',
    period=30)

In [ ]:
x_shuf1

0         1         2     ...      2997      2998      2999
87982  -0.088735 -0.104907 -0.132443  ...  0.027965  0.047633  0.040203
68528   0.600975  0.594189  0.584011  ...  0.065738  0.092458  0.064042
37536  -0.034046 -0.019160 -0.031151  ...  0.099106  0.117714  0.104481
49751  -0.057419  0.039738 -0.062918  ... -0.197653 -0.003799 -0.241190
55894  -0.095076 -0.096344 -0.092538  ...  0.062681  0.065219  0.076638
...          ...       ...       ...  ...       ...       ...       ...
66766   0.052937  0.026246  0.023744  ...  0.040008  0.028748  0.070036
42895  -0.049391  0.001550 -0.088372  ...  0.059136 -0.002879  0.044518
7912   -0.019850 -0.035156  0.010284  ...  0.091118  0.068638  0.093031
56020   0.194850  0.157920  0.165215  ... -0.003478  0.004273  0.003817
114355  0.089229 -0.028979  0.099336  ... -0.036010 -0.023706 -0.014917

[116718 rows x 3000 columns]

In [ ]:
model_30_epochs = model_b(verbose=VBS)
hist_30_epochs = model_30_epochs.fit(
    x_shuf1_rshp, y_shuf1_, batch_size=batch_size, epochs=30, verbose=VBS, callbacks=[cp_callback_erfan_model_30_epochs]
)

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                              

In [ ]:
y_pred_shuf1 = model_30_epochs.predict(pp_X_sub_482, batch_size=batch_size)
y_pred_shuf1 = np.array([np.argmax(s) for s in y_pred_shuf1])
f1_cnn_shuf1 = f1_score(y_sub_482, y_pred_shuf1, average="macro")
if VBS:
    print("F1 score: {}".format(f1_cnn_shuf1))
    report_shuf1 = classification_report(y_sub_482, y_pred_shuf1)
    print(report_shuf1)

F1 score: 0.7348806164895081
              precision    recall  f1-score   support

           0       0.98      0.96      0.97      1351
           1       0.38      0.12      0.19       187
           2       0.76      0.93      0.84       807
           3       0.91      0.84      0.87       318
           4       0.80      0.81      0.81       407

    accuracy                           0.87      3070
   macro avg       0.77      0.73      0.73      3070
weighted avg       0.86      0.87      0.85      3070



30 epochs

F1 score: 0.7348806164895081

              precision    recall  f1-score   support

           0       0.98      0.96      0.97      1351
           1       0.38      0.12      0.19       187
           2       0.76      0.93      0.84       807
           3       0.91      0.84      0.87       318
           4       0.80      0.81      0.81       407

    accuracy                           0.87      3070
    macro avg       0.77      0.73      0.73      3070
    weighted avg       0.86      0.87      0.85      3070


  
   


50 subjects

30 epochs: 87%

In [ ]:
latest_30_epochs=tf.train.latest_checkpoint(checkpoint_path_erfan_model_dir)

In [ ]:
print(latest_30_epochs)
epochs_30_model_path="/gdrive/My Drive/physionet/dt_shuf1_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf1_30_epochs.ckpt"

/gdrive/My Drive/physionet/dt_shuf1_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf1_30_epochs.ckpt


In [ ]:
loaded_model_30_epochs = model_b(verbose=VBS)
loaded_model_30_epochs.load_weights(epochs_30_model_path)

Model: "model_3"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                            

In [ ]:
optimizer = Adam(lr=1e-4)

In [ ]:
loaded_model_30_epochs.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
y_sub_482_=np_utils.to_categorical(y_sub_482)

In [ ]:
y_sub_482_

array([[1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       ...,
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.]], dtype=float32)

In [ ]:
accuracy_loaded_model_30_epochs = loaded_model_30_epochs.evaluate(pp_X_sub_482, y_sub_482_)

96/96 [==============================] - 2s 5ms/step - loss: 0.4133 - accuracy: 0.8678


In [ ]:
drive.flush_and_unmount()

In [ ]:
from google.colab import drive

drive.mount('gdrive', force_remount=True)

40 epochs

In [ ]:
checkpoint_path_erfan_model_40_epochs = "/content/gdrive/My Drive/physionet/dt_shuf1_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf1_40_epochs.ckpt"

In [ ]:
cp_callback_erfan_model_40_epochs = tf.keras.callbacks.ModelCheckpoint(
    filepath=checkpoint_path_erfan_model_40_epochs,
    monitor='accuracy', 
    verbose=1, 
    save_weights_only=True,
    mode = 'max',
    period=40)

In [ ]:
model_40_epochs = model_b(verbose=VBS)
hist_40_epochs = model_40_epochs.fit(
    x_shuf1_rshp, y_shuf1_, batch_size=batch_size, epochs=40, verbose=VBS, callbacks=[cp_callback_erfan_model_40_epochs]
)

Model: "model_2"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                            

In [ ]:
y_pred_shuf1_40_epochs = model_40_epochs.predict(pp_X_sub_482, batch_size=batch_size)
y_pred_shuf1_40_epochs = np.array([np.argmax(s) for s in y_pred_shuf1_40_epochs])
f1_cnn_shuf1 = f1_score(y_sub_482, y_pred_shuf1_40_epochs, average="macro")
if VBS:
    print("F1 score: {}".format(f1_cnn_shuf1))
    report_shuf1 = classification_report(y_sub_482, y_pred_shuf1_40_epochs)
    print(report_shuf1)

F1 score: 0.7459357322832272
              precision    recall  f1-score   support

           0       0.98      0.93      0.96      1351
           1       0.27      0.25      0.26       187
           2       0.79      0.92      0.85       807
           3       0.89      0.93      0.91       318
           4       0.82      0.69      0.75       407

    accuracy                           0.86      3070
   macro avg       0.75      0.74      0.75      3070
weighted avg       0.86      0.86      0.85      3070



50 subjects

40 epochs: 86%

In [ ]:
#print(latest_30_epochs)
epochs_40_model_path="/content/gdrive/My Drive/physionet/dt_shuf1_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf1_40_epochs.ckpt"

In [ ]:
loaded_model_40_epochs = model_b(verbose=VBS)
loaded_model_40_epochs.load_weights(epochs_40_model_path)

Model: "model_6"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                            

In [ ]:
loaded_model_40_epochs.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
accuracy_loaded_model_40_epochs = loaded_model_40_epochs.evaluate(pp_X_sub_482, y_sub_482_)

96/96 [==============================] - 3s 6ms/step - loss: 0.4413 - accuracy: 0.8554


In [ ]:
pip install -q -U keras-tuner

     |████████████████████████████████| 133 kB 15.4 MB/s 


In [ ]:
#Hyperparameter Tuning

In [ ]:
import keras_tuner as kt

In [ ]:
my_model = model_b(verbose=VBS)

Model: "model_1"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                            

In [ ]:
n_classes=5
use_sub_layer=False
use_rnn=True
verbose=False

In [ ]:
class MyHyperModel(kt.HyperModel):
    def build(self, hp):
      inputLayer = Input(shape=(3000, 1), name='inLayer')
      convFine = Conv1D(filters=64, kernel_size=int(Fs/2), strides=int(Fs/16), padding='same', activation='relu', name='fConv1')(inputLayer)
      convFine = MaxPool1D(pool_size=8, strides=8, name='fMaxP1')(convFine)
      convFine = Dropout(rate=0.5, name='fDrop1')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv2')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv3')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv4')(convFine)
      convFine = MaxPool1D(pool_size=4, strides=4, name='fMaxP2')(convFine)
      fineShape = convFine.get_shape()
      convFine = Flatten(name='fFlat1')(convFine)
    
      # network to learn coarse features
      convCoarse = Conv1D(filters=32, kernel_size=Fs*4, strides=int(Fs/2), padding='same', activation='relu', name='cConv1')(inputLayer)
      convCoarse = MaxPool1D(pool_size=4, strides=4, name='cMaxP1')(convCoarse)
      convCoarse = Dropout(rate=0.5, name='cDrop1')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv2')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv3')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv4')(convCoarse)
      convCoarse = MaxPool1D(pool_size=2, strides=2, name='cMaxP2')(convCoarse)
      coarseShape = convCoarse.get_shape()
      convCoarse = Flatten(name='cFlat1')(convCoarse)
    
      # concatenate coarse and fine cnns
      mergeLayer = concatenate([convFine, convCoarse], name='merge_1')
      outLayer = Dropout(rate=0.5, name='mDrop1')(mergeLayer)
    
      outLayer = Reshape((1, outLayer.get_shape()[1]), name='reshape1')(outLayer)
      outLayer = LSTM(64, return_sequences=True)(outLayer)
      outLayer = LSTM(64, return_sequences=False)(outLayer)

      # Classify
      outLayer = Dense(n_classes, activation='softmax', name='outLayer')(outLayer)
      model = Model(inputLayer, outLayer)
      optimizer = Adam(lr=1e-4)
      model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['acc'])
      #model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
      if verbose:
        model.summary()
      return model

    def fit(self, hp, model, *args, **kwargs):
        return model.fit(
            x_shuf1_rshp,y_shuf1_,
            epochs=hp.Choice("epochs", [10, 20, 30, 40, 50, 60], default=30)
        )

In [ ]:
tuner = kt.RandomSearch(
    MyHyperModel(),
    objective="loss",
    max_trials=3,
    overwrite=True,
    #directory="my_dir",
    #project_name="tune_hypermodel",
)

In [ ]:
tuner.search(x_shuf1_rshp,y_shuf1_)

Trial 3 Complete [00h 15m 58s]
loss: 0.3501404821872711

Best loss So Far: 0.34688031673431396
Total elapsed time: 00h 43m 04s
INFO:tensorflow:Oracle triggered exit


In [ ]:
tuner.search_space_summary()

Search space summary
Default search space size: 1
epochs (Choice)
{'default': 30, 'conditions': [], 'values': [10, 20, 30, 40, 50, 60], 'ordered': True}


In [ ]:
tuner.results_summary()

Results summary
Results in ./untitled_project
Showing 10 best trials
Trial summary
Hyperparameters:
epochs: 30
Score: 0.34688031673431396
Trial summary
Hyperparameters:
epochs: 30
Score: 0.3501404821872711
Trial summary
Hyperparameters:
epochs: 20
Score: 0.39059850573539734


In [ ]:
class MyHyperModel(kt.HyperModel):
    def build(self, hp):
      inputLayer = Input(shape=(3000, 1), name='inLayer')
      convFine = Conv1D(filters=64, kernel_size=int(Fs/2), strides=int(Fs/16), padding='same', activation='relu', name='fConv1')(inputLayer)
      convFine = MaxPool1D(pool_size=8, strides=8, name='fMaxP1')(convFine)
      convFine = Dropout(rate=0.5, name='fDrop1')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv2')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv3')(convFine)
      convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv4')(convFine)
      convFine = MaxPool1D(pool_size=4, strides=4, name='fMaxP2')(convFine)
      fineShape = convFine.get_shape()
      convFine = Flatten(name='fFlat1')(convFine)
    
      # network to learn coarse features
      convCoarse = Conv1D(filters=32, kernel_size=Fs*4, strides=int(Fs/2), padding='same', activation='relu', name='cConv1')(inputLayer)
      convCoarse = MaxPool1D(pool_size=4, strides=4, name='cMaxP1')(convCoarse)
      convCoarse = Dropout(rate=0.5, name='cDrop1')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv2')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv3')(convCoarse)
      convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv4')(convCoarse)
      convCoarse = MaxPool1D(pool_size=2, strides=2, name='cMaxP2')(convCoarse)
      coarseShape = convCoarse.get_shape()
      convCoarse = Flatten(name='cFlat1')(convCoarse)
    
      # concatenate coarse and fine cnns
      mergeLayer = concatenate([convFine, convCoarse], name='merge_1')
      outLayer = Dropout(rate=0.5, name='mDrop1')(mergeLayer)
    
      outLayer = Reshape((1, outLayer.get_shape()[1]), name='reshape1')(outLayer)
      outLayer = LSTM(64, return_sequences=True)(outLayer)
      outLayer = LSTM(64, return_sequences=False)(outLayer)

      # Classify
      outLayer = Dense(n_classes, activation='softmax', name='outLayer')(outLayer)
      model = Model(inputLayer, outLayer)
      optimizer = Adam(lr=1e-4)
      model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['acc'])
      #model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
      if verbose:
        model.summary()
      return model

    def fit(self, hp, model, *args, **kwargs):
        return model.fit(
            x_shuf1_rshp,y_shuf1_,
            epochs=hp.Choice("epochs", [10, 20, 30, 40, 50, 60])
        )

In [ ]:
Fs=100

In [ ]:
tuner = kt.RandomSearch(
    MyHyperModel(),
    objective="loss",
    max_trials=3,
    overwrite=True,
    #directory="my_dir",
    #project_name="tune_hypermodel",
)

In [ ]:
tuner.search(x_shuf1_rshp,y_shuf1_)

Trial 3 Complete [00h 22m 22s]
loss: 0.26555460691452026

Best loss So Far: 0.26555460691452026
Total elapsed time: 00h 40m 11s
INFO:tensorflow:Oracle triggered exit


In [ ]:
tuner.results_summary()

Results summary
Results in ./untitled_project
Showing 10 best trials
Trial summary
Hyperparameters:
epochs: 50
Score: 0.26555460691452026
Trial summary
Hyperparameters:
epochs: 30
Score: 0.3435036242008209
Trial summary
Hyperparameters:
epochs: 10
Score: 0.4447125494480133


In [ ]:
#tuner.get_best_hyperparameters()[0]

In [ ]:
#best_model = tuner.get_best_models(1)[0]